### PR TITLE
[ha-agent] Run HA enabled integrations only on leader agent

### PR DESCRIPTION
**Pull request by AlexandreYang**
### What does this PR do?

[ha-agent] Run agent integration only on leader agent

This PR changes a lot of files, but the actual change is only few lines located in pkg/collector/worker/worker.go:
https://github.com/DataDog/datadog-agent/pull/31186/files#diff-fa837122b5056960e5f1bcbe5e6da5ed11701b8e75a2dff72681905d9a1e0314R150-R234

the rest of the changes are only about injecting correctly the component.

Marked as `qa/rc-required` since it's a new feature that have impact on whether integrations skip or not.

- Manual QA should test that when HA Agent feature is enabled:
  - non-HA integrations always run (whether the Agent is leader or not)
  - HA integrations are only run on leader agent (not run on follower agents)
- We plan to have a e2e test soonish for HA Agent feature (NDMII-3190)

### Motivation

Needed for HA Agent feature.

### Describe how to test/QA your changes

1/ Run 2 Agents with HA Agent feature enabled:
```
ha_agent:
  enabled: true
  group: my-group01
```

2/ Write messages from backend to HA_AGENT RC product

4/ Check that only leader agent is running snmp integrations

### CODE REVIEW

File: pkg/collector/worker/worker.go [hunk 1]
```diff
@@ -141,6 +147,14 @@ func (w *Worker) Run() {
 			continue
 		}
 
+		if !w.shouldRunIntegrationInstance(check) {
+			// TODO: TEST ME
```
Comment (jedupau) [inline pkg/collector/worker/worker.go:151]: is it tested ? 👀

**Review by jedupau — APPROVED** [review]

File: pkg/collector/worker/worker.go [hunk 2]
```diff
@@ -141,6 +147,13 @@ func (w *Worker) Run() {
 			continue
 		}
 
+		if !w.shouldRunIntegrationInstance(check) {
+			checkLogger.Debug("HA Integration skipped")
```
Comment (vickenty) [inline pkg/collector/worker/worker.go:151]: Could you please add an explanation about why the check is skipped, in the similar fashion to how it is done on line 146?

File: pkg/collector/worker/worker.go [hunk 3]
```diff
@@ -141,6 +147,13 @@ func (w *Worker) Run() {
 			continue
 		}
 
+		if !w.shouldRunIntegrationInstance(check) {
```
Comment (vickenty) [inline pkg/collector/worker/worker.go:150]: Would it be possible to move this block before `checksTracker.AddCheck` block directly above? Then you wouldn't need to call `DeleteCheck` here, and we don't need to deduplicate runs of non-running check.

File: pkg/collector/worker/worker.go [hunk 4]
```diff
@@ -212,6 +225,13 @@ func (w *Worker) Run() {
 	log.Debugf("Runner %d, worker %d: Finished processing checks.", w.runnerID, w.ID)
 }
 
+func (w *Worker) shouldRunIntegrationInstance(check check.Check) bool {
+	if w.haAgent.Enabled() && w.haAgent.IsHaIntegration(check.String()) {
+		return w.haAgent.IsLeader()
+	}
+	return true
```
Comment (vickenty) [inline pkg/collector/worker/worker.go:232]: Would it make sense to encapsulate this logic in the HA agent component?

```suggestion
	return w.haAgent.ShouldRun(check)
```

File: releasenotes/notes/NDMII-3154-ha-agent-collector-worker-22f3972469c669c3.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    [ha-agent] Run HA enabled integrations only on leader agent
```
Comment (jhgilbert) [inline releasenotes/notes/NDMII-3154-ha-agent-collector-worker-22f3972469c669c3.yaml:11]:
```suggestion
    [ha-agent] Run HA enabled integrations only on leader Agent
```

**Review by jhgilbert — APPROVED** [review]
Approved with a minor suggestion for style, thanks!

File: pkg/collector/worker/worker.go (hunk 4 above)
Comment (AlexandreYang) [reply pkg/collector/worker/worker.go]: Good idea, addressed here: https://github.com/DataDog/datadog-agent/pull/31186/commits/5ed60bbfe99f9fa395e9f6cd87f830836a7d29a9

File: pkg/collector/worker/worker.go (hunk 3 above)
Comment (AlexandreYang) [reply pkg/collector/worker/worker.go]: Good idea, fixed here https://github.com/DataDog/datadog-agent/pull/31186/commits/5ed60bbfe99f9fa395e9f6cd87f830836a7d29a9

File: comp/haagent/impl/haagent.go [hunk 6]
```diff
@@ -50,6 +50,15 @@ func (h *haAgentImpl) SetLeader(leaderAgentHostname string) {
 	h.isLeader.Store(agentHostname == leaderAgentHostname)
 }
 
+// ShouldRunIntegration return true if the agent integrations is should to run.
+// When ha-agent is disabled, the agent behave as standalone agent (non HA) and will always run all integrations.
+func (h *haAgentImpl) ShouldRunIntegration(integrationName string) bool {
+	if h.Enabled() {
+		return validHaIntegrations[integrationName] && h.isLeader.Load()
```
Comment (vickenty) [inline comp/haagent/impl/haagent.go:57]: I don't think this logic is correct: when I run the agent with `DD_HA_AGENT_ENABLED=true`, the agent skips all checks:
```
2024-11-27 16:26:36 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:76 in Debug) | check:uptime | Check is an HA integration and current agent is not leader, skipping execution...
2024-11-27 16:26:36 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:76 in Debug) | check:ntp | Check is an HA integration and current agent is not leader, skipping execution...
2024-11-27 16:26:36 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:76 in Debug) | check:service_discovery | Check is an HA integration and current agent is not leader, skipping execution...
2024-11-27 16:26:37 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:76 in Debug) | check:memory | Check is an HA integration and current agent is not leader, skipping execution...
2024-11-27 16:26:38 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:76 in Debug) | check:io | Check is an HA integration and current agent is not leader, skipping execution...
```

Could you also please add a test that makes sure that non-HA checks continue to run when HA-agent is enabled.

**Review by vickenty — CHANGES_REQUESTED** [review]

File: comp/haagent/impl/haagent.go (hunk 6 above)
Comment (AlexandreYang) [reply comp/haagent/impl/haagent.go]: @vickenty thanks for catching that, it's fixed here https://github.com/DataDog/datadog-agent/pull/31186/commits/f65d6d34acdfbf2a3c2e21091264c4b565d0e949 with updated tests.

looks like I introduced the bug when I moved move the logic from worker.go to haagent component: https://github.com/DataDog/datadog-agent/pull/31186/commits/5ed60bbfe99f9fa395e9f6cd87f830836a7d29a9#diff-fa837122b5056960e5f1bcbe5e6da5ed11701b8e75a2dff72681905d9a1e0314R225-R226

**Review by vickenty — APPROVED** [review]

**Pull request merged**